suro-kafka-producer/src/main/java/com/netflix/suro/sink/kafka/KafkaSink.java

```diff
@@ -105,34 +119,155 @@ private void setServoReporter() {
                     long hashCode = keyField.hashCode();
-                    partition = Math.abs((int)(hashCode ^ (hashCode >>> 32))) % numPartitions;
+                    partition = Math.abs((int) (hashCode ^ (hashCode >>> 32))) % numPartitions;
                 }
             } catch (Exception e) {
                 log.error("Exception on getting key field: " + e.getMessage());
             }
         }
 
-        producer.send(new ProducerRecord(message.getRoutingKey(), partition, null, message.getMessage().getPayload()));
+        try {
+            producer.send(
+                new ProducerRecord(message.getRoutingKey(), partition, null, message.getMessage().getPayload()),
+                new Callback() {
+                    @Override
+                    public void onCompletion(RecordMetadata metadata, Exception e) {
+                        if (e != null) {
+                            log.error("Exception while sending", e);
+                            DynamicCounter.increment(
+                                MonitorConfig
+                                    .builder("failedRecord")
+                                    .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                                    .build());
+                            droppedRecords.incrementAndGet();
+                            runRecordCounterListener();
+                        } else {
+                            DynamicCounter.increment(
+                                MonitorConfig
+                                    .builder("sentRecord")
+                                    .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                                    .build());
+                            sentRecords.incrementAndGet();
+                            runRecordCounterListener();
+                        }
+                    }
+                });
+        } catch (Exception e) {
+            log.error("Exception before sending", e);
+            dropMessage(message);
+        }
+    }
+
+    private void dropMessage(MessageContainer message) {
+        DynamicCounter.increment(
+            MonitorConfig
+                .builder("droppedRecord")
+                .withTag(TagKey.ROUTING_KEY, message.getRoutingKey())
+                .build());
+        droppedRecords.incrementAndGet();
+        runRecordCounterListener();
     }
 
     @Override
     public void open() {
         producer = new KafkaProducer(props);
+        subscription = stream
+            .filter(new Func1<MessageContainer, Boolean>() {
+                @Override
+                public Boolean call(MessageContainer message) {
+                    if (!metadataWaitingQueuePolicy.acquire()) {
+                        dropMessage(message);
+                        return false;
+                    } else {
+                        return true;
+                    }
+                }
+            })
+            .subscribe(
```

Comment:
Instead of manually creating Runnable and submitting to executor, we can use `Observable.subscribeOn()` and pass in an rx Scheduler, which can be created from an executor.